Sources/CartonCLI/Commands/Init.swift

```diff
@@ -25,43 +25,85 @@ struct Init: ParsableCommand {
 
   func run() throws {
     let terminal = InteractiveWriter.stdout
 
-    guard let name = name ?? localFileSystem.currentWorkingDirectory?.basename else {
-      terminal.write("Project name could not be inferred\n", inColor: .red)
-      return
-    }
-    guard let currentDir = localFileSystem.currentWorkingDirectory else {
-      terminal.write("Failed to get current working directory.\n", inColor: .red)
-      return
+    guard let currentDirectory = localFileSystem.currentWorkingDirectory else {
+      throw ProjectInitialisationError.cannotFindCurrentDirectory
     }
+
+    /// The kind of template the project is to be created with.
+    /// - Note: This value is of the type `Templates`, not `Template`.
     let template = self.template ?? .basic
+
+    let projectName = name ?? currentDirectory.basename
+    let projectPath = name == nil ? currentDirectory : currentDirectory.appending(component: projectName)
+    
+    // FIXME: Replace with `NSString.abbreviatingWithTildeInPath`?
+    let packagePathRelativeToHomeDirectory = projectPath.relative(to: localFileSystem.homeDirectory)
+
     terminal.write("Creating new project with template ")
     terminal.write("\(template.rawValue)", inColor: .green)
     terminal.write(" in ")
-    terminal.write("\(name)\n", inColor: .cyan)
+    terminal.write("~/\(packagePathRelativeToHomeDirectory.pathString)\n", inColor: .cyan)
 
-    guard let packagePath = self.name == nil ?
-      localFileSystem.currentWorkingDirectory :
-      AbsolutePath(name, relativeTo: currentDir)
-    else {
-      terminal.write("Path to project could be created.\n", inColor: .red)
-      return
+    do {
+      // `LocalFileSystem.createDirectory` wraps `FileManager.createDirectory`, which throws an NSError.
+      try localFileSystem.createDirectory(projectPath)
+    } catch let nsError as NSError {
+        throw ProjectInitialisationError.cannotCreateDirectory(path: packagePathRelativeToHomeDirectory, reason: nsError.localizedDescription)
+    }
+    
+    // FIXME: Handle this error-catching in each `Template.create()` instead.
+    do {
+      // FIXME:  The first `template` is of type `Templates`, and two consecutive `template` looks odd.
+      // Maybe rename `Templates` as `TemplateKind`, and refactor `Template` so the template doesn't have to be created this way.
+      try template.template.createProject(
+        at: projectPath,
+        inSubdirectory: name != nil,
+        on: localFileSystem,
+        terminal: terminal
+      )
+    } catch let nsError as NSError {
+        throw ProjectInitialisationError.cannotCreateWithTemplate(template, reason: nsError.localizedDescription)
+    } catch let error {
+        throw ProjectInitialisationError.cannotCreateWithTemplate(template, reason: "\(error)")
     }
-    try localFileSystem.createDirectory(packagePath)
-    try template.template.create(
-      on: localFileSystem,
-      project: .init(name: name, path: packagePath, inPlace: self.name == nil),
```

Comment:
I changed this function from taking a `Project` to taking a project's path as an argument. Because there is some weird error of `Project`'s synthesized member-wise initializer having `internal` access level, although `Project` is declared `public`.

Reply:
That's the default behavior in Swift for `structs` as far as I know: member-wise initializer's are always `internal` even in `public` types, to get a public one you have to declare it explicitly.